chalk-ir/src/lib.rs

```diff
@@ -327,11 +328,148 @@ pub struct Ty<I: Interner> {
+        }
+        LifetimeData::Placeholder(_) => {
+            TypeFlags::HAS_RE_PLACEHOLDER
+                | TypeFlags::HAS_FREE_LOCAL_REGIONS
+                | TypeFlags::HAS_FREE_REGIONS
+        }
+        LifetimeData::Static | LifetimeData::Phantom(_, _) | LifetimeData::BoundVar(_) => {
+            TypeFlags::empty()
+        }
+    }
+}
+
+/// Compute type flags for Substitution<I>
+fn compute_substitution_flags<I: Interner>(
+    substitution: &Substitution<I>,
+    interner: &I,
+) -> TypeFlags {
+    let mut flags = TypeFlags::empty();
+    for generic_arg in substitution.iter(&interner) {
+        flags |= compute_generic_arg_flags(generic_arg, &interner);
+    }
+    flags
+}
+
+/// Compute type flags for GenericArg<I>
+fn compute_generic_arg_flags<I: Interner>(generic_arg: &GenericArg<I>, interner: &I) -> TypeFlags {
+    match generic_arg.data(&interner) {
+        GenericArgData::Ty(ty) => ty.data(interner).flags,
+        GenericArgData::Lifetime(lifetime) => compute_lifetime_flags(lifetime, interner),
+        GenericArgData::Const(constant) => {
+            let data = constant.data(&interner);
+            let flags = data.ty.data(interner).flags;
+            match data.value {
+                ConstValue::BoundVar(_) => flags,
+                ConstValue::InferenceVar(_) => {
+                    flags | TypeFlags::HAS_CT_INFER | TypeFlags::STILL_FURTHER_SPECIALIZABLE
+                }
+                ConstValue::Placeholder(_) => {
+                    flags | TypeFlags::HAS_CT_PLACEHOLDER | TypeFlags::STILL_FURTHER_SPECIALIZABLE
+                }
+                ConstValue::Concrete(_) => flags,
+            }
+        }
+    }
+}
+
+/// Compute type flags for aliases
+fn compute_alias_flags<I: Interner>(alias_ty: &AliasTy<I>, interner: &I) -> TypeFlags {
+    match alias_ty {
+        AliasTy::Projection(projection_ty) => {
+            TypeFlags::HAS_TY_PROJECTION
+                | compute_substitution_flags(&(projection_ty.substitution), interner)
+        }
+        AliasTy::Opaque(opaque_ty) => {
+            TypeFlags::HAS_TY_OPAQUE
+                | compute_substitution_flags(&(opaque_ty.substitution), interner)
+        }
+    }
+}
+
+/// Compute type flags for a TyKind
+fn compute_flags<I: Interner>(kind: &TyKind<I>, interner: &I) -> TypeFlags {
+    match kind {
+        TyKind::Adt(_, substitution)
+        | TyKind::AssociatedType(_, substitution)
+        | TyKind::Tuple(_, substitution)
+        | TyKind::Closure(_, substitution)
+        | TyKind::Generator(_, substitution)
+        | TyKind::GeneratorWitness(_, substitution)
+        | TyKind::FnDef(_, substitution) => compute_substitution_flags(substitution, interner),
+        TyKind::Scalar(_) | TyKind::Str | TyKind::Never | TyKind::Foreign(_) => TypeFlags::empty(),
+        TyKind::OpaqueType(_, substitution) => {
+            TypeFlags::HAS_TY_OPAQUE | compute_substitution_flags(substitution, interner)
```

Comment:
So, looking through how this is used in rustc  I think that the majority of code would not expect this it to be set for `TyKind::OpaqueType`, only `AliasTy::Opaque`.